File: DailyChallenges/CPP/LeetCode_875.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        auto check = [&](int k){
            int _h = 0;
            for (int p: piles){
                _h += (p+k-1)/k;
                if (_h > h) return true;
            }
            return false;
        };
        int l = 1, r = 1e9;
        while (l < r){
            int m = l + (r-l)/2;
            if (check(m)) l = m+1; else r=m;
        }
        return l;
    }
};

File: DailyChallenges/Scala/LeetCode_875.scala
object Solution {
    def minEatingSpeed(piles: Array[Int], h: Int): Int = {
        def check(k: Int): Boolean = {
            piles.foldLeft(0)((s, p) => {
                val _s = s + (p+k-1)/k
                if (_s > h) return true
                _s
            })
            false
        }

        def bs(l: Int, r: Int): Int = if (l>=r) l else {
            val m = l + (r-l)/2
            if (check(m)) bs(m+1,r) else bs(l, m)
        }

        bs(1, 1e9.toInt)
    }
}
